### ISSUE TITLE
Infer type/version/loader from packwiz

### ISSUE THREAD

**Issue by lhvy** [itzg/docker-minecraft-server#1818]
### Enhancement Type

Improve an existing feature

### Describe the enhancement

Currently using packwiz with Quilt for the latest release of Minecraft and the Quilt loader can cause problems because the container sees that the type is set to Quilt, and downloads whichever version of the loader it prefers, which in this case is actually 0.18.x beta, which is incompatible with mods in my packwiz pack. This could be fixed if the version, type and loader version env vars were simply replaced with values read directly from the packwiz pack.

**Comment by itzg** [comment]
I am gradually working towards this type of capability with CurseForge modpacks. So, this enhancement makes sense for packwiz also. 

The unique challenge with the packwiz version is that it'll need to pre-download the definition and interpret just enough to get the versions.

**Comment by itzg** [comment]
[As discussed in Discord](https://discord.com/channels/660567679458869252/660569641550217327/1104122137175920740) this track nicely with TYPE=AUTO_CURSEFORGE where it would be TYPE=PACKWIZ

**Comment by WilsontheWolf** [comment]
Had a similar issue, where despite my pack having the minecraft version set, it downloaded the latest and then it couldn't boot due to mod incompatibilities.

**Comment by jcotton42** [comment]
> The unique challenge with the packwiz version is that it'll need to pre-download the definition and interpret just enough to get the versions.

This seems like it shouldn't be *too* bad? _knocks on wood_

The `pack.toml` pointed to by the URL contains all the pertinent versions under the `[versions]` heading.

**Comment by jcotton42** [comment]
One way that comes to mind is, if the TYPE is packwiz, then set the VERSION, TYPE, and other variables from the manifest, then let the image continue as normal.

e.g. a packwiz manifest with
```toml
[versions]
minecraft = "1.21.1"
neoforge = "21.1.169"
```

Would result in VERSION=1.12.1, TYPE=NEOFORGE, and NEOFORGE_VERSION=21.1.169. The rest of the startup scripts would then continue as before, none the wiser that their settings were coming from packwiz instead of the container env.

**Comment by jcotton42** [comment]
I threw together a proof of concept supporting just NeoForge https://github.com/itzg/docker-minecraft-server/compare/master...jcotton42:docker-minecraft-server:versions-from-packwiz

Running `docker run -e TYPE=packwiz -e PACKWIZ_URL=https://github.com
/jcotton42/modpacks/raw/refs/heads/main/packs/DragonPack/pack.toml -e EULA=TRUE minecraft-server` where pack.toml is this file

https://github.com/jcotton42/modpacks/blob/55b2f63ad887dbf0f30b7e166cd00a742fc1bea4/packs/DragonPack/pack.toml#L1-L16

Results in
```
[init] Running as uid=1000 gid=1000 with /data as 'drwxr-x--- 2 1000 1000 4096 May 20 05:59 /data'
[init] Image info: buildtime=local,version=local,revision=local
[init] Resolving type given NEOFORGE
[mc-image-helper] 06:05:59.423 INFO  : Running NeoForge 21.1.169 installer for Minecraft 1.21.1. This might take a while...
[init] Running packwiz installer against URL: https://github.com/jcotton42/modpacks/raw/refs/heads/main/packs/DragonPack/pack.toml
packwiz-installer was started without packwiz-installer-bootstrap. Use the bootstrapper for automatic updates! (Disregard this message if you have your own update mechanism)
Loading manifest file...
Loading pack file...
Loading MultiMC pack file...
MultiMC not detected
Checking local files...
Modpack name: Dragon Pack
Checking local files...
Comparing new files...
Side changed, invalidating all mods
Validating existing files...
...
```

So it seems to work.

**Comment by itzg** [comment]
Thanks @jcotton42 ! 

I would have preferred it to be implemented in mc-image-helper like [CurseForge](https://github.com/itzg/mc-image-helper/tree/master/src/main/java/me/itzg/helpers/curseforge) and [Modrinth](https://github.com/itzg/mc-image-helper/tree/master/src/main/java/me/itzg/helpers/modrinth) packs; however, the proposed approach seems good. 

FYI mc-image-helper has a `toml-path` subcommand which can be used to extract values from a toml file rather than needing to use awk. Refer to [the usage section](https://github.com/itzg/mc-image-helper/tree/master#usage). 

A PR would be great if you want to proceed.

**Comment by jcotton42** [comment]
I could definitely put it in image-helper if you would prefer.

What kind of approach were you envisioning? Just bootstrap the necessary Minecraft and modloader versions, leaving the actual pack installations for the later script parts? Or something else?

**Comment by rhullah** [comment]
I love that movement is happening with this, thanks @jcotton42!!

**Comment by itzg** [comment]
> I could definitely put it in image-helper if you would prefer.

Yeah it would be great if you could take a stab at enhancing it there. 

> What kind of approach were you envisioning? Just bootstrap the necessary Minecraft and modloader versions, leaving the actual pack installations for the later script parts? Or something else?

Ideally the image helper approach could track Modrinth and CurseForge incorporating the packwiz installer download execution from

https://github.com/itzg/docker-minecraft-server/blob/1370f58c267dcd936b50acfc3d44b77159ca9bb1/scripts/start-setupModpack#L30-L47

and directly kicking off the modloader install as part of it. 

Perhaps a larger exercise than you originally planned so a script-side approach is good too.

**Comment by rhullah** [comment]
I know I'm not the one offering to do the work, but what did you mean by this? 

> track Modrinth and CurseForge

Wouldn't image helper not care where it got the mods it would only care about which mod loader and version to use/install, right?

**Comment by itzg** [comment]
That's a great question. To (hopefully) clarify, the CurseForge and Modrinth modpack installers in mc-image-helper do the following all in one invocation (but via various resuable bits of code):

1. Obtain the modpack metadata
2. Install the mods, etc from that
3. Determine the modloader and version
4. Download the modloader installer (Forge, Fabric, etc)
5. Run the modloader installer

For packwiz I'm thinking it could do all the following in one invocation:

1. Obtain the packwiz toml (metadata)
2. Obtain the packwiz installer, which is actually using an mc-image-helper subcommand already
3. Run the packwiz installer
4. Continue with steps 3-5 above

**Comment by jcotton42** [comment]
Lemme check my understanding.

A new script, deployPackwiz, that checks options passed through the environment (currently just that PACKWIZ_URL isn't empty, maybe also a force reinstall flag?), then invokes the image-helper.

The image-helper would then download pack.toml from the URL, parsing out the Minecraft and (if applicable) modloader versions.

Then it would either
1. Install just Minecraft and the modloader
2. Install Minecraft and the modloader, then run the bootstrapper

If (1), the bootstrapper would be run by the scripts, either in deployPackwiz, or where it already exists in setupModpack (if in deployPackwiz, the second will need to be suppressed).

Some other things that came to mind
- Since we're parsing the pack.toml, the version number of the pack could be cached, allowing skipping the bootstrapper run on subsequent server starts if the modpack version hasn't changed (we'll want an override flag in this case though)
- How does this container currently handle packwiz being unable to download a Curseforge mod? My guess is it currently does not, and can't really given that would require somehow parsing the output of packwiz

**Comment by itzg** [comment]
Quick note: the packwiz installer only leaves behind packwiz.json, but it just contains checksum info.

> Edit: this comment was drafted before your most recent reply itzg
> 
> Lemme check my understanding.
> 
> A new script, deployPackwiz, that checks options passed through the environment (currently just that PACKWIZ_URL isn't empty, maybe also a force reinstall flag?), then invokes the image-helper.

Yeah, I'm thinking a non-empty PACKWIZ_URL is still the trigger to kick off the new image helper subcommand.

> The image-helper would then download pack.toml from the URL, parsing out the Minecraft and (if applicable) modloader versions.
> 
> Then it would either
> 
> 1. Install just Minecraft and the modloader
> 2. Install Minecraft and the modloader, then run the bootstrapper

As you saw in my concurrent comment :), option 2 is what I'm thinking if by "bootstrapper" you mean the packwiz installer.

> If (1), the bootstrapper would be run by the scripts, either in deployPackwiz, or where it already exists in setupModpack (if in deployPackwiz, the second will need to be suppressed).

I'm thinking introduce a new, but very small `start-deployPackwiz' much like https://github.com/itzg/docker-minecraft-server/blob/1370f58c267dcd936b50acfc3d44b77159ca9bb1/scripts/start-deployModrinth. Follow the pattern there to let the startup sequence proceed as normal from there; however in this new layout the packwiz bit in start-setupModpack would be gone.

> Some other things that came to mind
> 
> * Since we're parsing the pack.toml, the version number of the pack could be cached, allowing skipping the bootstrapper run on subsequent server starts if the modpack version hasn't changed (we'll want an override flag in this case though)

Yeah, it'll be great to follow the "manifest" logic like

https://github.com/itzg/mc-image-helper/blob/60f6f998072369e8a279474bfddebe8fdcd8fac0/src/main/java/me/itzg/helpers/modrinth/InstallModrinthModpackCommand.java#L130-L133 

> * How does this container currently handle packwiz being unable to download a Curseforge mod? My guess is it currently does not, and can't really given that would require somehow parsing the output of packwiz

I'm not sure that packwiz can download from CurseForge. Honestly I haven't dug deep into their support other than knowing enough to kick off their installer :)

**Comment by itzg** [comment]
> A new script, deployPackwiz

Sorry, I had missed that you already said what I said :)

**Comment by rhullah** [comment]
Packwiz can install mods from Curseforge, Modrinth, a direct file URL, and I think even from GitHub Releases (this may be new).

**Comment by jcotton42** [comment]
> I'm not sure that packwiz can download from CurseForge.

It does, though the API key in the bootstrapper is unfortunately hardcoded at the moment. Much like this container, it is blocked if 3rd-party downloads are turned off for that mod. I think at the moment it just tells you which mods failed and their project URLs, and that you'll need to download them manually.

I'm not sure how that would be fixed on our end, aside from ditching the bootstrapper for our own downloader (which shouldn't actually be too bad, could be future work?) that would use the CF repo when rejected.

**Comment by jcotton42** [comment]
> Packwiz can install mods from Curseforge, Modrinth, a direct file URL, and I think even from GitHub Releases (this may be new).

You can also drop files directly in the pack. And you're correct about GitHub, though at download time it's handled the same way as a direct URL, the GitHub metadata is only used for `packwiz update`.

**Comment by itzg** [comment]
> It does, though the API key in the bootstrapper is unfortunately hardcoded at the moment. 

Wow, they hardcode the API key? That's "generous" (or something) of them. I sure wouldn't distribute my API key with my installer. 😄 

> Much like this container, it is blocked if 3rd-party downloads are turned off for that mod. 

That's no longer an issue with my CurseForge installer (in mc-image-helper).

> I think at the moment it just tells you which mods failed and their project URLs, and that you'll need to download them manually.
> 
> I'm not sure how that would be fixed on our end, aside from ditching the bootstrapper for our own downloader (which shouldn't actually be too bad, could be future work?) that would use the CF repo when rejected.

If that's the case, then that should be an enhancement request for Packwiz IMHO.

**Comment by jcotton42** [comment]
Anyhow, I'll take a stab at doing this "properly" this week.

**Comment by jcotton42** [comment]
Hm, speaking of manifests @itzg, I see that they're supposed to keep track of files.

I'm not sure how to do that with packwiz though, when we're having their installer handle that. Would anything break if I left the files entry in the manifest blank?

**Comment by itzg** [comment]
Good point. Yes that files list can be left empty.

**Comment by ryleu** [comment]
> Anyhow, I'll take a stab at doing this "properly" this week.

would you mind linking the PR you open here once you start on it? would love to follow development / lend a hand if possible

**Comment by jcotton42** [comment]
> > Anyhow, I'll take a stab at doing this "properly" this week.
> 
> would you mind linking the PR you open here once you start on it? would love to follow development / lend a hand if possible

I've actually already started it and am a good way into it, just got sidetracked and then Deltarune came out. I'll get the PR up this weekend, even if not complete (unless @itzg prefers I wait to have something complete before opening the PR).

**Comment by itzg** [comment]
Feel free to kick off the PR as soon as you're ready especially since you can always start it out in draft mode.

**Comment by jcotton42** [comment]
> > Anyhow, I'll take a stab at doing this "properly" this week.
> 
> would you mind linking the PR you open here once you start on it? would love to follow development / lend a hand if possible

@ryleu https://github.com/itzg/mc-image-helper/pull/577

**Comment by jcotton42** [comment]
So, I got busy with life stuff soon after opening the PR. I can resume this weekend if this feature still isn't present.

**Comment by itzg** [comment]
Thanks @jcotton42 , that would still be very helpful. 

FWIW I've had various life stuff going on and hadn't even realized time had passed as much.

**Comment by jcotton42** [comment]
So my last comment turned out to be an unintentional lie, sorry about that. I'm getting close, but hit an interesting point: packwiz doesn't require declaring a modloader, in which case we would want to just do a vnailla install. However, all the logic for downloading the vanilla server jar is in the scripts in docker-minecraft-server, not the image helper. Should we move the vanilla logic to image helper?

**Comment by itzg** [comment]
> Should we move the vanilla logic to image helper?

Yeah, I actually had wanted to move that over at some point. Just to clarify, that would be good in a separate PR.

**Comment by jcotton42** [comment]
Sounds good. That should be simple enough that I can do it before finishing packwiz.

**Comment by itzg** [comment]
Yeah the hardest part would have been the vanilla version resolution. But that's already implemented in mc image helper.

**Comment by jcotton42** [comment]
~~Done with the mc-image-helper part of the vanilla bit, but now I'm unsure how I'm supposed to get my local copy cleanly into the docker image for testing. The Dockerfile seems to be hardcoded from the web.~~

~~I guess I could pass in a `file://` URL for the `MC_HELPER_BASE_URL` arg? Or bind mount my build into the container? Both are clunky and I'm wondering how you do it.~~

Never mind, I didn't look at the docs hard enough https://docker-minecraft-server.readthedocs.io/en/latest/misc/contributing/development/

**Comment by jcotton42** [comment]
I have opened itzg/mc-image-helper#812 and #4165 for the required vanilla functionality.

**Comment by jcotton42** [comment]
Back to the actual packwiz stuff now, looking for some advice for designing for testability. In https://github.com/jcotton42/mc-image-helper/blob/b19cfb84d2e01d5fa81e71b32e524ee6b926f6e4/src/main/java/me/itzg/helpers/packwiz/InstallPackwizModpackCommand.java#L132 right now I just have the installers being created and called directly, but that gets messy in testing.

My first thought is maybe an interface, we'll call it `ModLoaderInstaller` that has `installNeoforge`, `installForge`, `installFabric`, `installQuilt` methods. The production version calls the live installers, the testing ones do fakes. I'll get started on that idea, but I'd love suggestions.

(Edit: it seems this is basically what the Modrinth pack installer does, so I guess I was on the money)

Speaking of which, do we want packwiz setups to support all the various advanced modloader options. Such as the ability to override the promotions and maven repo URLs for installing Forge? That could make the CLI pretty messy.

**Comment by itzg** [comment]
> (Edit: it seems this is basically what the Modrinth pack installer does, so I guess I was on the money)

I have to admit that overall I've been pretty sloppy with my testing of the modpack handlers and such. Modrinth was the first one where I started to address that more consistently, so yep, that would be a nice pattern to follow. 

> Speaking of which, do we want packwiz setups to support all the various advanced modloader options. Such as the ability to override the promotions and maven repo URLs for installing Forge? That could make the CLI pretty messy.

Packwiz does things like alternate maven repo URLs? That sounds excessive. I'm not sure if I want to encourage/support such things, but if that's what is normal for packwiz then it'll have to be supported to some extent. I really have no exposure to packwiz -- why does it seem to be so radically different and complicated compared to the typical modpack formats like CurseForge and Modrinth? 

Perhaps there's a middle ground where the installer class can be parameterized further, but not necessarily expose those via the respective CLI.

**Comment by jcotton42** [comment]
> Packwiz does things like alternate maven repo URLs?

I was referring to stuff like ForgeUrlArgs in this project's forge installer. Packwiz itself does not install your modloader, it just needs to know so it can auto select the right file when you do `packwiz curseforge add` or `packwiz modrinth add` with just a project URL (URLs to specific files are added without checks).

At bootstrap time it just downloads exactly what's in the pack and doesn't even look at the configured Minecraft or modloader versions.

**Comment by jcotton42** [comment]
The bulk of packwiz's complexity is really there for authoring time, to make mods easy to add and update while ensuring they are compatible with your game and modloaders. At install time the bootstrapper just has to recursively go through the metafiles and download the URLs within to the specified filename (except for CurseForge mods, because those don't have stable download URLs).

**Comment by ryleu** [comment]
i have used packwiz for years and never touched the custom maven feature (didn't even know it existed). i feel like that's a follow up on this PR if someone is so inclined

**Comment by jcotton42** [comment]
Packwiz does not have custom maven features. I was referring to ForgeUrlArgs.

**Comment by itzg** [comment]
Oh didn't realize you meant the other way around. That URL customization is mainly there for unit test mocking and for this one time that Forge had an outage and I temporarily needed to let users point to an alternative.
